Disallow the `-fchar8_t` and `-fno-char8_t` options for C modes

Clang 15 trunk currently allows the `-fchar8_t` and `-fno-char8_t` options to be specified when compiling in C modes and the former has the effect of causing `char8_t` to be both a keyword and a type specifier. This behavior is not desirable; in C23, `char8_t` is a typedef of `unsigned char` declared in `uchar.h`. See [N2653](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2653.htm) for details.

```
$ cat t.c
char8_t c8;

$ clang --version
clang version 15.0.0 (https://github.com/tahonermann/llvm-project.git 5531abaf7158652bf7411937781780e9f6358ecf)
...

$ clang -c t.c
t.c:1:1: error: unknown type name 'char8_t'
char8_t c8;
^
1 error generated.

$ clang -c -fchar8_t t.c
<no error, but we want one!>
```

This issue will track changing Clang to reject the `-fchar8_t` and `-fno-char8_t` options in C modes or to ignore them with an appropriate diagnostic.

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -317,7 +317,7 @@ Modified Compiler Flags
 
 - The ``-mexecute-only`` and ``-mpure-code`` flags are now accepted for AArch64 targets. (#GH125688)
 
-- The ``-fchar8_t`` flag is no longer consider in non-C++ languages modes. (#GH55373)
+- The ``-fchar8_t`` flag is no longer considered in non-C++ languages modes. (#GH55373)
 
 Removed Compiler Flags
 -------------------------

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -317,6 +317,8 @@ Modified Compiler Flags
 
 - The ``-mexecute-only`` and ``-mpure-code`` flags are now accepted for AArch64 targets. (#GH125688)
 
+- The ``-fchar8_t`` flag is no longer consider in non-C++ languages modes. (#GH55373)
+
 Removed Compiler Flags
 -------------------------
 

--- a/clang/include/clang/Driver/Options.td
+++ b/clang/include/clang/Driver/Options.td
@@ -3879,7 +3879,8 @@ defm char8__t : BoolFOption<"char8_t",
   LangOpts<"Char8">, Default<cpp20.KeyPath>,
   PosFlag<SetTrue, [], [ClangOption], "Enable">,
   NegFlag<SetFalse, [], [ClangOption], "Disable">,
-  BothFlags<[], [ClangOption, CC1Option], " C++ builtin type char8_t">>;
+  BothFlags<[], [ClangOption, CC1Option], " C++ builtin type char8_t">>,
+  ShouldParseIf<cplusplus.KeyPath>;
 def fshort_wchar : Flag<["-"], "fshort-wchar">, Group<f_Group>,
   HelpText<"Force wchar_t to be a short unsigned int">;
 def fno_short_wchar : Flag<["-"], "fno-short-wchar">, Group<f_Group>,

--- a/clang/test/Lexer/char8_t.cpp
+++ b/clang/test/Lexer/char8_t.cpp
@@ -5,6 +5,7 @@
 // RUN: %clang_cc1 -std=c++17 -verify %s
 // RUN: %clang_cc1 -std=c++17 -verify %s -fno-char8_t
 // RUN: %clang_cc1 -std=c++20 -verify %s -fno-char8_t
+// RUN: %clang_cc1 -x c -verify %s -fchar8_t
 
 #if defined(__cpp_char8_t) != defined(CHAR8_T)
 #error wrong setting for __cpp_char_t

